substrate/network/src/protocol.rs

```diff
@@ -221,15 +293,21 @@ impl<B: BlockT> Protocol<B> where
 	/// Called by peer when it is disconnecting
 	pub fn on_peer_disconnected(&self, io: &mut SyncIo, peer: PeerId) {
 		trace!(target: "sync", "Disconnecting {}: {}", peer, io.peer_info(peer));
+
+		// lock all the the peer lists so that add/remove peer events are in order
+		let mut sync = self.sync.write();
+		let mut spec = self.specialization.write();
+
 		let removed = {
-			let mut peers = self.peers.write();
+			let mut peers = self.context_data.peers.write();
```

Comment:
Docs for `Protocol` says that any locks must be taken in order of declaration of fields. Seems like in this place this rule is violated.

Reply:
The code behaves as if `context_data.peers` is located after `specialization`. I guess it should be fine if we just reorder fields, I guess.

(However, the idea of touching locks inside other objects sounds not plausible for me).